docs/DafnyRef/Grammar.md

```diff
@@ -562,6 +562,52 @@ but the names of the parameters are optional. This is used in algebraic
 datatype definitions.
 
 ### 2.6.5. Quantifier Domains {#sec-quantifier-domains}
+
+Several Dafny constructs bind one or more variables to a range of possible values.
+For example, the quantifier `forall x: nat | x <= 5 :: x * x <= 25` has the meaning
+"for all integers x between 0 and 5, the square of x is at most 25".
+Similarly, the set comprehension `set x: nat | x <= 5 :: f(x)` can be read as
+"the set containing the result of applying f to x, for each integer x between 0 and 5".
+The common syntax that specifies the bound variables and what values they take on
+is known as the *quantifier domain*; in the previous examples this is `x: nat | x <= 5`, 
+which binds the variable `x` to the values `0`, `1`, `2`, `3`, `4`, and `5`.
+
+Here are some more examples.
+
+- `x: byte` (where `type byte = x | 0 <= x < 256`)
+- `x: nat | x <= 5`
+- `x <- integerSet`
+- `x: nat <- integerSet`
+- `x: nat <- integerSet | x % 2 == 0`
+- `x: nat, y: nat | x < 2 && y < 2`
+- `i | 0 <= i < |s|, y <- s[i] | i < y`
+
+A quantifier domain declares one or more *quantified variables*, separated by commas.
+Each variable declaration can be nothing more than a variable name, but it 
+may also include any of three optional elements:
+
+1. The optional syntax `: T` declares the type of the quantified variable.
+   If not provided, it will be inferred from context.
+
+2. The optional syntax `| E` attaches a boolean expression `E` as a *quantified variable range*,
+   which restricts the bindings to values that satisfy this expression.
+   In the example above `x <= 5` is the range attached to the `x` variable declaration.
+
+3. The optional syntax `<- C` attaches a collection expression `C` as a *quantified variable domain*.
```

Comment:
Could you move this one before 2, so that it's the order in which we parse things?

Reply:
I was on the edge of doing this already and you have tipped me over :)